segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/nfs/NFSStorage.java

```diff
@@ -0,0 +1,443 @@
+import io.pravega.segmentstore.storage.Storage;
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.RandomAccessFile;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.channels.ClosedChannelException;
+import java.nio.channels.FileChannel;
+import java.nio.channels.NonWritableChannelException;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.file.AccessDeniedException;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.nio.file.attribute.FileAttribute;
+import java.nio.file.attribute.PosixFileAttributes;
+import java.nio.file.attribute.PosixFilePermission;
+import java.nio.file.attribute.PosixFilePermissions;
+import java.time.Duration;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+
+import static java.nio.file.attribute.PosixFilePermission.OWNER_WRITE;
+
+/**
+ * Storage adapter for file system based Tier2.
+ *
+ * Each segment is represented as a single file on the underlying storage. As the writes at an offset for a POSIX
+ * file is idempotent, there is no need for locking when a container fails over to another host.
+ * As Pravega does not modify data in Tier2 once written even a contention in writing will cause same data to be
+ * written at the same offset till the time the original host gives up ownership.
+ */
+
+@Slf4j
+public class NFSStorage implements Storage {
+
+    //region members
+
+    private final NFSStorageConfig config;
+    private final ExecutorService executor;
+
+    //endregion
+
+    //region constructor
+
+    public NFSStorage(NFSStorageConfig config, ExecutorService executor) {
+        Preconditions.checkNotNull(config, "config");
+        Preconditions.checkNotNull(executor, "executor");
+        this.config = config;
+        this.executor = executor;
+    }
+
+    //endregion
+
+    //region Storage implementation
+    @Override
+    public void initialize(long containerEpoch) {
+
+    }
+
+    @Override
+    public CompletableFuture<SegmentHandle> openRead(String streamSegmentName) {
+        final CompletableFuture<SegmentHandle> retVal = new CompletableFuture<>();
+
+        executor.execute( () -> {
+            syncOpenRead(streamSegmentName, retVal);
+        });
+
+        return retVal;
+    }
+
+
+    @Override
+    public CompletableFuture<Integer> read(SegmentHandle handle, long offset, byte[] buffer, int bufferOffset, int
+            length, Duration timeout) {
+        final CompletableFuture<Integer> retVal = new CompletableFuture<>();
+
+        executor.execute( () -> {
+         syncRead(handle, offset, buffer, bufferOffset, length, timeout, retVal);
```

Comment:
It looks like your indentation is off in the next couple of methods, this is using 2 spaces, and the next 2 use 3 spaces.